File: DataMaker.Forms.Renderer.WordPress/assets/fn.js
/**
 * Dm — browser-side runtime for the DataMaker `Fn` expression library.
 * JsCompiler emits `Dm.<Name>(args)` calls; this file provides the bodies.
 * Function names and arities mirror DataMaker.Expressions.Functions.Fn so
 * source maps 1-to-1.
 *
 * Semantics notes:
 *  - String comparisons that are OrdinalIgnoreCase in C# use `.toLowerCase()`
 *    equality here. Adequate for ASCII form values; pathological unicode
 *    cases will diverge from the server.
 *  - Regex Match/Capture/Replace use the `i` flag to match C#'s IgnoreCase.
 *  - Date inputs accept anything `Date(...)` parses; outputs are YYYY-MM-DD
 *    so they round-trip through DataMaker date fields.
 */
(function (global) {
  'use strict';

  var Dm = {};

  function isEq(a, b) {
    return String(a).toLowerCase() === String(b).toLowerCase();
  }

  function toStr(v) { return v == null ? '' : String(v); }
  function toArr(v) { return Array.isArray(v) ? v : []; }

  // ── Decompose / Compose ──────────────────────────────────────────
  Dm.Split = function (input, separator) {
    if (input == null || separator == null) return [];
    var parts = String(input).split(String(separator));
    return parts.filter(function (s) { return s.length > 0; });
  };
  Dm.Join     = function (arr, sep)    { return toArr(arr).join(toStr(sep)); };
  Dm.Concat   = function (a, b)        { return toStr(a) + toStr(b); };
  Dm.Concat3  = function (a, b, c)     { return toStr(a) + toStr(b) + toStr(c); };
  Dm.Concat4  = function (a, b, c, d)  { return toStr(a) + toStr(b) + toStr(c) + toStr(d); };

  // ── Select ───────────────────────────────────────────────────────
  Dm.At = function (arr, index) {
    var a = toArr(arr);
    var i = index | 0;
    if (i < 0) i = a.length + i;
    return (i >= 0 && i < a.length) ? a[i] : '';
  };
  Dm.First = function (arr) { var a = toArr(arr); return a.length ? a[0] : ''; };
  Dm.Last  = function (arr) { var a = toArr(arr); return a.length ? a[a.length - 1] : ''; };

  Dm.Take = function (arr, n) {
    var a = toArr(arr); return a.slice(0, Math.min(n | 0, a.length));
  };
  Dm.Skip = function (arr, n) {
    var a = toArr(arr); var k = n | 0;
    return k >= a.length ? [] : a.slice(k);
  };
  Dm.Slice = function (arr, start, count) {
    var a = toArr(arr); var s = start | 0; var c = count | 0;
    if (s < 0) s = Math.max(0, a.length + s);
    if (s >= a.length) return [];
    return a.slice(s, Math.min(s + c, a.length));
  };

  Dm.IndexOf = function (arr, value) {
    var a = toArr(arr); var v = toStr(value);
    for (var i = 0; i < a.length; i++) if (isEq(a[i], v)) return i;
    return -1;
  };
  Dm.Where  = function (arr, value) {
    var v = toStr(value);
    return toArr(arr).filter(function (s) { return isEq(s, v); });
  };
  Dm.Except = function (arr, value) {
    var v = toStr(value);
    return toArr(arr).filter(function (s) { return !isEq(s, v); });
  };

  // ── Branch ───────────────────────────────────────────────────────
  Dm.If        = function (cond, then, otherwise) { return cond ? then : otherwise; };
  Dm.Coalesce  = function (a, b)                  { return toStr(a).length > 0 ? a : b; };
  Dm.Coalesce3 = function (a, b, c) {
    if (toStr(a).length > 0) return a;
    if (toStr(b).length > 0) return b;
    return c;
  };
  Dm.Switch = function (value, case1, result1, def) {
    return isEq(value, case1) ? result1 : def;
  };
  Dm.Switch2 = function (value, c1, r1, c2, r2, def) {
    if (isEq(value, c1)) return r1;
    if (isEq(value, c2)) return r2;
    return def;
  };
  Dm.Switch3 = function (value, c1, r1, c2, r2, c3, r3, def) {
    if (isEq(value, c1)) return r1;
    if (isEq(value, c2)) return r2;
    if (isEq(value, c3)) return r3;
    return def;
  };

  // ── Convert ──────────────────────────────────────────────────────
  Dm.ToInt   = function (s) { var n = parseInt(toStr(s), 10); return Number.isFinite(n) ? n : 0; };
  Dm.ToLong  = function (s) { var n = parseInt(toStr(s), 10); return Number.isFinite(n) ? n : 0; };
  Dm.IntToStr0 = function (n) { return String(n | 0); };
  Dm.IntToStr = function (n, format) {
    var v = n | 0;
    var fmt = toStr(format).trim();
    if (!fmt) return String(v);
    var head = fmt.charAt(0).toUpperCase();
    var width = parseInt(fmt.slice(1), 10);
    if (head === 'D') {
      // D<n>: zero-pad to n digits, preserve sign.
      var sign = v < 0 ? '-' : '';
      var body = Math.abs(v).toString();
      if (Number.isFinite(width) && width > body.length) body = '0'.repeat(width - body.length) + body;
      return sign + body;
    }
    if (head === 'N') {
      // N0..N<n>: thousands separator, n decimal places.
      var digits = Number.isFinite(width) ? Math.max(0, width) : 0;
      return v.toLocaleString(undefined, { minimumFractionDigits: digits, maximumFractionDigits: digits });
    }
    if (head === 'F') {
      var d = Number.isFinite(width) ? Math.max(0, width) : 2;
      return v.toFixed(d);
    }
    return String(v);
  };

  // ── Array transforms ─────────────────────────────────────────────
  Dm.Insert = function (arr, index, value) {
    var a = toArr(arr).slice();
    var i = Math.max(0, Math.min(index | 0, a.length));
    a.splice(i, 0, value);
    return a;
  };
  Dm.Remove = function (arr, index) {
    var a = toArr(arr).slice();
    var i = index | 0;
    if (i < 0) i = a.length + i;
    if (i < 0 || i >= a.length) return a;
    a.splice(i, 1);
    return a;
  };
  Dm.Filter = function (arr, value) {
    var v = toStr(value);
    return toArr(arr).filter(function (s) { return !isEq(s, v); });
  };
  Dm.Reverse = function (arr) { return toArr(arr).slice().reverse(); };
  Dm.Sort = function (arr) {
    return toArr(arr).slice().sort(function (a, b) {
      var la = String(a).toLowerCase(), lb = String(b).toLowerCase();
      return la < lb ? -1 : la > lb ? 1 : 0;
    });
  };
  Dm.Distinct = function (arr) {
    var a = toArr(arr); var seen = {}; var out = [];
    for (var i = 0; i < a.length; i++) {
      var k = String(a[i]).toLowerCase();
      if (!seen.hasOwnProperty(k)) { seen[k] = true; out.push(a[i]); }
    }
    return out;
  };
  Dm.Map = function (arr, prefix, suffix) {
    var p = toStr(prefix); var s = toStr(suffix);
    return toArr(arr).map(function (x) { return p + toStr(x) + s; });
  };
  Dm.Union = function (a, b) {
    return Dm.Distinct(toArr(a).concat(toArr(b)));
  };
  Dm.Intersect = function (a, b) {
    var setB = {}; toArr(b).forEach(function (x) { setB[String(x).toLowerCase()] = true; });
    return Dm.Distinct(toArr(a).filter(function (x) { return setB[String(x).toLowerCase()]; }));
  };
  Dm.SetExcept = function (a, b) {
    var setB = {}; toArr(b).forEach(function (x) { setB[String(x).toLowerCase()] = true; });
    return Dm.Distinct(toArr(a).filter(function (x) { return !setB[String(x).toLowerCase()]; }));
  };

  // ── String transforms ────────────────────────────────────────────
  Dm.Lower = function (s) { return toStr(s).toLowerCase(); };
  Dm.Upper = function (s) { return toStr(s).toUpperCase(); };
  Dm.Trim  = function (s) { return toStr(s).trim(); };
  Dm.TitleCase = function (s) {
    return toStr(s).toLowerCase().replace(/(^|\s|[\-_\/])([a-z])/g, function (_, sep, ch) {
      return sep + ch.toUpperCase();
    });
  };
  Dm.Pad = function (s, width, padChar) {
    var v = toStr(s); var w = width | 0;
    if (v.length >= w) return v;
    var c = toStr(padChar).charAt(0) || ' ';
    return c.repeat(w - v.length) + v;
  };
  Dm.PadRight = function (s, width, padChar) {
    var v = toStr(s); var w = width | 0;
    if (v.length >= w) return v;
    var c = toStr(padChar).charAt(0) || ' ';
    return v + c.repeat(w - v.length);
  };
  Dm.Left  = function (s, n) { var v = toStr(s); return v.slice(0, Math.min(n | 0, v.length)); };
  Dm.Right = function (s, n) {
    var v = toStr(s); var k = n | 0;
    return k >= v.length ? v : v.slice(v.length - k);
  };

  // ── Regex ────────────────────────────────────────────────────────
  function compileRegex(pattern) {
    try { return new RegExp(toStr(pattern), 'i'); } catch (_) { return null; }
  }
  Dm.Match = function (input, pattern) {
    var re = compileRegex(pattern); if (!re) return false;
    return re.test(toStr(input));
  };
  Dm.Capture = function (input, pattern, group) {
    var re = compileRegex(pattern); if (!re) return '';
    var m = toStr(input).match(re);
    var g = group | 0;
    return (m && g < m.length) ? (m[g] || '') : '';
  };
  Dm.Replace = function (input, pattern, replacement) {
    try {
      var re = new RegExp(toStr(pattern), 'gi');
      return toStr(input).replace(re, toStr(replacement));
    } catch (_) { return toStr(input); }
  };

  // ── Aggregates ───────────────────────────────────────────────────
  Dm.Len    = function (arr) { return toArr(arr).length; };
  Dm.LenStr = function (s)   { return toStr(s).length; };
  Dm.Any = function (arr, value) {
    var v = toStr(value); return toArr(arr).some(function (x) { return isEq(x, v); });
  };
  Dm.All = function (arr, value) {
    var v = toStr(value); var a = toArr(arr);
    return a.length > 0 && a.every(function (x) { return isEq(x, v); });
  };
  Dm.CountOf = function (arr, value) {
    var v = toStr(value);
    return toArr(arr).reduce(function (n, x) { return n + (isEq(x, v) ? 1 : 0); }, 0);
  };
  Dm.Min  = function (a, b) { return Math.min(a, b); };
  Dm.Max  = function (a, b) { return Math.max(a, b); };
  Dm.MinD = function (a, b) { return Math.min(+a, +b); };
  Dm.MaxD = function (a, b) { return Math.max(+a, +b); };

  // ── Date operations ──────────────────────────────────────────────
  function parseDate(s) {
    if (s == null) return null;
    var d = new Date(String(s));
    return isNaN(d.getTime()) ? null : d;
  }
  function pad2(n) { n = n | 0; return (n < 10 ? '0' : '') + n; }
  function fmtYmd(d) {
    return d.getUTCFullYear() + '-' + pad2(d.getUTCMonth() + 1) + '-' + pad2(d.getUTCDate());
  }
  Dm.Year  = function (s) { var d = parseDate(s); return d ? d.getUTCFullYear() : 0; };
  Dm.Month = function (s) { var d = parseDate(s); return d ? d.getUTCMonth() + 1 : 0; };
  Dm.Day   = function (s) { var d = parseDate(s); return d ? d.getUTCDate() : 0; };
  Dm.AddDays = function (s, days) {
    var d = parseDate(s); if (!d) return toStr(s);
    d.setUTCDate(d.getUTCDate() + (days | 0));
    return fmtYmd(d);
  };
  Dm.DiffDays = function (a, b) {
    var d1 = parseDate(a), d2 = parseDate(b);
    if (!d1 || !d2) return 0;
    return Math.trunc((d1 - d2) / 86400000);
  };
  Dm.Today = function () { return fmtYmd(new Date()); };
  Dm.Now   = function () { return new Date().toISOString(); };

  global.Dm = Dm;
})(typeof window !== 'undefined' ? window : globalThis);

File: DataMaker.Expressions/Functions/Fn.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace DataMaker.Expressions.Functions;

/// <summary>
/// Expression function library for Data Maker calculated fields and query
/// expressions. Seeded from FileOrganizer's Fn library (project_expression_gaps),
/// adapted for form-field domain (path-specific helpers removed; keep array,
/// string, date, branch, convert, aggregate tiers).
///
/// Tier 1 — Primitives
///   Split, Concat, At, First, Last, Take, Skip, Slice, IndexOf, Where, Except,
///   If, Coalesce, Switch, ToInt, ToLong, IntToStr
/// Tier 2 — Transforms
///   Insert, Remove, Filter, Reverse, Sort, Distinct, Map, Union, Intersect, SetExcept,
///   Lower, Upper, TitleCase, Trim, Pad, PadRight, Left, Right, Match, Capture, Replace,
///   Len, LenStr, Any, All, CountOf, Min, Max,
///   Year, Month, Day, AddDays, DiffDays
/// </summary>
public static class Fn
{
    // ── Decompose ───────────────────────────────────────────────────

    public static string[] Split(string input, string separator)
        => input.Split(separator, StringSplitOptions.RemoveEmptyEntries);

    // ── Compose ─────────────────────────────────────────────────────

    public static string Join(string[] parts, string separator)
        => string.Join(separator, parts);

    public static string Concat(string a, string b) => a + b;
    public static string Concat3(string a, string b, string c) => a + b + c;
    public static string Concat4(string a, string b, string c, string d) => a + b + c + d;

    // ── Select ──────────────────────────────────────────────────────

    public static string At(string[] array, int index) => AtInternal(array, index);
    public static string First(string[] array) => array.Length > 0 ? array[0] : "";
    public static string Last(string[] array) => array.Length > 0 ? array[^1] : "";

    public static string[] Take(string[] array, int n)
        => array[..Math.Min(n, array.Length)];

    public static string[] Skip(string[] array, int n)
        => n >= array.Length ? [] : array[n..];

    public static string[] Slice(string[] array, int start, int count)
    {
        if (start < 0) start = Math.Max(0, array.Length + start);
        if (start >= array.Length) return [];
        return array[start..Math.Min(start + count, array.Length)];
    }

    public static int IndexOf(string[] array, string value)
        => Array.FindIndex(array, s => s.Equals(value, StringComparison.OrdinalIgnoreCase));

    public static string[] Where(string[] array, string value)
        => array.Where(s => s.Equals(value, StringComparison.OrdinalIgnoreCase)).ToArray();

    public static string[] Except(string[] array, string value)
        => array.Where(s => !s.Equals(value, StringComparison.OrdinalIgnoreCase)).ToArray();

    // ── Branch ──────────────────────────────────────────────────────

    public static string If(bool condition, string then, string otherwise) => condition ? then : otherwise;

    public static string Coalesce(string a, string b) => a.Length > 0 ? a : b;
    public static string Coalesce3(string a, string b, string c)
        => a.Length > 0 ? a : b.Length > 0 ? b : c;

    public static string Switch(string value, string case1, string result1, string defaultResult)
        => value.Equals(case1, StringComparison.OrdinalIgnoreCase) ? result1 : defaultResult;

    public static string Switch2(string value,
        string case1, string result1, string case2, string result2, string defaultResult)
        => value.Equals(case1, StringComparison.OrdinalIgnoreCase) ? result1
        : value.Equals(case2, StringComparison.OrdinalIgnoreCase) ? result2
        : defaultResult;

    public static string Switch3(string value,
        string case1, string result1, string case2, string result2,
        string case3, string result3, string defaultResult)
        => value.Equals(case1, StringComparison.OrdinalIgnoreCase) ? result1
        : value.Equals(case2, StringComparison.OrdinalIgnoreCase) ? result2
        : value.Equals(case3, StringComparison.OrdinalIgnoreCase) ? result3
        : defaultResult;

    // ── Convert ─────────────────────────────────────────────────────

    public static int ToInt(string s) => int.TryParse(s, out var n) ? n : 0;
    public static long ToLong(string s) => long.TryParse(s, out var n) ? n : 0;
    public static string IntToStr(int n, string format) => n.ToString(format);
    public static string IntToStr0(int n) => n.ToString();

    // ── Array transforms ────────────────────────────────────────────

    public static string[] Insert(string[] array, int index, string value)
    {
        var list = new List<string>(array);
        list.Insert(Math.Clamp(index, 0, list.Count), value);
        return list.ToArray();
    }

    public static string[] Remove(string[] array, int index)
    {
        if (index < 0) index = array.Length + index;
        if (index < 0 || index >= array.Length) return array;
        var list = new List<string>(array);
        list.RemoveAt(index);
        return list.ToArray();
    }

    public static string[] Filter(string[] array, string value)
        => array.Where(s => !s.Equals(value, StringComparison.OrdinalIgnoreCase)).ToArray();

    public static string[] Reverse(string[] array)
    { var copy = (string[])array.Clone(); Array.Reverse(copy); return copy; }

    public static string[] Sort(string[] array)
    { var copy = (string[])array.Clone(); Array.Sort(copy, StringComparer.OrdinalIgnoreCase); return copy; }

    public static string[] Distinct(string[] array)
        => array.Distinct(StringComparer.OrdinalIgnoreCase).ToArray();

    public static string[] Map(string[] array, string prefix, string suffix)
        => array.Select(s => prefix + s + suffix).ToArray();

    public static string[] Union(string[] a, string[] b)
        => a.Union(b, StringComparer.OrdinalIgnoreCase).ToArray();

    public static string[] Intersect(string[] a, string[] b)
        => a.Intersect(b, StringComparer.OrdinalIgnoreCase).ToArray();

    public static string[] SetExcept(string[] a, string[] b)
        => a.Except(b, StringComparer.OrdinalIgnoreCase).ToArray();

    // ── String transforms ───────────────────────────────────────────

    public static string Lower(string s) => s.ToLowerInvariant();
    public static string Upper(string s) => s.ToUpperInvariant();
    public static string Trim(string s) => s.Trim();

    public static string TitleCase(string s)
        => CultureInfo.InvariantCulture.TextInfo.ToTitleCase(s.ToLowerInvariant());

    public static string Pad(string s, int width, string padChar)
        => s.PadLeft(width, padChar.Length > 0 ? padChar[0] : ' ');

    public static string PadRight(string s, int width, string padChar)
        => s.PadRight(width, padChar.Length > 0 ? padChar[0] : ' ');

    public static string Left(string s, int n) => s[..Math.Min(n, s.Length)];
    public static string Right(string s, int n) => n >= s.Length ? s : s[^n..];

    // ── Regex ───────────────────────────────────────────────────────

    public static bool Match(string input, string pattern)
        => Regex.IsMatch(input, pattern, RegexOptions.IgnoreCase);

    public static string Capture(string input, string pattern, int group)
    {
        var m = Regex.Match(input, pattern, RegexOptions.IgnoreCase);
        return m.Success && group < m.Groups.Count ? m.Groups[group].Value : "";
    }

    public static string Replace(string input, string pattern, string replacement)
        => Regex.Replace(input, pattern, replacement, RegexOptions.IgnoreCase);

    // ── Aggregates ──────────────────────────────────────────────────

    public static int Len(string[] array) => array.Length;
    public static int LenStr(string s) => s.Length;

    public static bool Any(string[] array, string value)
        => array.Any(s => s.Equals(value, StringComparison.OrdinalIgnoreCase));

    public static bool All(string[] array, string value)
        => array.All(s => s.Equals(value, StringComparison.OrdinalIgnoreCase));

    public static int CountOf(string[] array, string value)
        => array.Count(s => s.Equals(value, StringComparison.OrdinalIgnoreCase));

    public static int Min(int a, int b) => Math.Min(a, b);
    public static int Max(int a, int b) => Math.Max(a, b);
    public static decimal MinD(decimal a, decimal b) => Math.Min(a, b);
    public static decimal MaxD(decimal a, decimal b) => Math.Max(a, b);

    // ── Date operations ─────────────────────────────────────────────

    public static int Year(string dateStr) => TryParseDate(dateStr, out var d) ? d.Year : 0;
    public static int Month(string dateStr) => TryParseDate(dateStr, out var d) ? d.Month : 0;
    public static int Day(string dateStr) => TryParseDate(dateStr, out var d) ? d.Day : 0;

    public static string AddDays(string dateStr, int days)
        => TryParseDate(dateStr, out var d) ? d.AddDays(days).ToString("yyyy-MM-dd") : dateStr;

    public static int DiffDays(string dateStr1, string dateStr2)
        => TryParseDate(dateStr1, out var d1) && TryParseDate(dateStr2, out var d2)
            ? (int)(d1 - d2).TotalDays : 0;

    public static string Today() => DateTime.UtcNow.ToString("yyyy-MM-dd");
    public static string Now() => DateTime.UtcNow.ToString("O");

    // ── Internal helpers ────────────────────────────────────────────

    private static string AtInternal(string[] array, int index)
    {
        if (index < 0) index = array.Length + index;
        return index >= 0 && index < array.Length ? array[index] : "";
    }

    private static bool TryParseDate(string s, out DateTime result)
        => DateTime.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.None, out result);
}
